### PR TITLE
Conformance between spec and API: knn

### PR DESCRIPTION
**Pull request by amgrigoriev**

### CODE REVIEW

File: cpp/oneapi/dal/algo/knn/backend/cpu/infer_kernel_brute_force_classification.cpp [hunk 1]
```diff
@@ -23,17 +23,17 @@
...
-struct infer_kernel_cpu<Float, method::brute_force> {
-    infer_result operator()(const context_cpu &ctx,
-                            const descriptor_base &desc,
-                            const infer_input &input) const {
+template <typename Float, typename Task>
+struct infer_kernel_cpu<Float, method::brute_force, Task> {
+    infer_result<Task> operator()(const context_cpu &ctx,
+                                  const descriptor_base<Task> &desc,
+                                  const infer_input<Task> &input) const {
         throw unimplemented("k-NN brute force method is not implemented for CPU");
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/knn/backend/cpu/infer_kernel_brute_force_classification.cpp:31]: Is it correct?

File: cpp/oneapi/dal/algo/knn/detail/infer_ops_dpc.cpp [hunk 2]
```diff
@@ -21,24 +21,24 @@
...
+struct ONEAPI_DAL_EXPORT infer_ops_dispatcher<data_parallel_policy, Float, Method, Task> {
+    infer_result<Task> operator()(const data_parallel_policy& ctx,
+                                  const descriptor_base<Task>& params,
+                                  const infer_input<Task>& input) const {
         using kernel_dispatcher_t =
-            dal::backend::kernel_dispatcher<backend::infer_kernel_cpu<Float, Method>,
-                                            backend::infer_kernel_gpu<Float, Method>>;
+            dal::backend::kernel_dispatcher<backend::infer_kernel_cpu<Float, Method, Task>,
+                                            backend::infer_kernel_gpu<Float, Method, Task>>;
         return kernel_dispatcher_t{}(ctx, params, input);
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/knn/detail/infer_ops_dpc.cpp:32]: Please use one style `()` or `{}`. It's not a critical remark

**Review by KulikovNikita — APPROVED** [review]

File: cpp/oneapi/dal/algo/knn/detail/infer_ops_dpc.cpp (hunk 2 above)
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/knn/detail/infer_ops_dpc.cpp:32]:
```suggestion
        return kernel_dispatcher_t()(ctx, params, input);
```

File: examples/oneapi/cpp/source/knn/knn_cls_kd_tree_dense_batch.cpp [hunk 3]
```diff
@@ -34,9 +34,7 @@ int main(int argc, char const *argv[]) {
   const auto y_train = dal::read<dal::table>(dal::csv::data_source{train_label_file_name});
 
   const auto knn_desc =
-      dal::knn::descriptor<float, oneapi::dal::knn::method::kd_tree>()
-          .set_class_count(5)
-          .set_neighbor_count(1);
+      dal::knn::descriptor<float, oneapi::dal::knn::method::kd_tree, oneapi::dal::knn::task::classification>(5, 1);
```
Comment (mchernov-intel) [inline examples/oneapi/cpp/source/knn/knn_cls_kd_tree_dense_batch.cpp:37]: may be better to save methods? I mean `set_class_count` and `set_class_count`? It looks more representative.

File: examples/oneapi/dpc/source/knn/knn_cls_brute_force_dense_batch.cpp [hunk 4]
```diff
@@ -33,9 +33,7 @@ void run(sycl::queue& queue) {
     const auto y_train = dal::read<dal::table>(queue, dal::csv::data_source{train_label_file_name});
 
     const auto knn_desc =
-        dal::knn::descriptor<float, oneapi::dal::knn::method::brute_force>()
-            .set_class_count(5)
-            .set_neighbor_count(1);
+        dal::knn::descriptor<float, oneapi::dal::knn::method::brute_force, oneapi::dal::knn::task::classification>(5, 1);
```
Comment (mchernov-intel) [inline examples/oneapi/dpc/source/knn/knn_cls_brute_force_dense_batch.cpp:36]: same

File: cpp/oneapi/dal/algo/knn/common.hpp [hunk 5]
```diff
@@ -48,36 +59,44 @@ class ONEAPI_DAL_EXPORT descriptor_base : public base {
...
+template <typename Float = descriptor_base<task::by_default>::float_t,
+          typename Method = descriptor_base<task::by_default>::method_t,
+          typename Task = task::by_default>
+class descriptor : public descriptor_base<Task> {
 public:
     using tag_t = detail::tag;
     using float_t = Float;
     using method_t = Method;
 
+    explicit descriptor(std::int64_t class_count, std::int64_t neighbor_count) {
```
Comment (mchernov-intel) [inline cpp/oneapi/dal/algo/knn/common.hpp:74]: Have we already approved such public constructors?

**Review by mchernov-intel — APPROVED** [review]

File: cpp/oneapi/dal/algo/knn/backend/cpu/infer_kernel_brute_force_classification.cpp [hunk 6]
```diff
@@ -18,20 +18,22 @@
 #include "oneapi/dal/backend/interop/error_converter.hpp"
 #include "oneapi/dal/detail/common.hpp"
 
+#define DAAL_SYCL_INTERFACE
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/knn/backend/cpu/infer_kernel_brute_force_classification.cpp:21]: This should not be defined here

File: cpp/oneapi/dal/algo/knn/detail/model_impl.hpp [hunk 7]
```diff
@@ -33,4 +34,4 @@ class detail::model_impl : public base {
     interop_model* interop_;
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/knn/detail/model_impl.hpp:34]: There is no delete for interop model. It looks like memory leak. That's what I've found in KD-tree method training:
```
auto interop = new daal_interop_model_t(model_ptr);
const auto model_impl = std::make_shared<detail::model_impl>(interop);
```

File: cpp/oneapi/dal/algo/knn/common.hpp [hunk 8]
```diff
@@ -48,36 +59,44 @@ class ONEAPI_DAL_EXPORT descriptor_base : public base {
     void set_class_count_impl(std::int64_t value);
     void set_neighbor_count_impl(std::int64_t value);
 
-    dal::detail::pimpl<detail::descriptor_impl> impl_;
+    dal::detail::pimpl<detail::descriptor_impl<Task>> impl_;
```
Comment (mchernov-intel) [inline cpp/oneapi/dal/algo/knn/common.hpp:62]: use `task_t` instead of Task

File: cpp/oneapi/dal/algo/knn/common.hpp [hunk 9]
```diff
@@ -48,36 +59,44 @@ class ONEAPI_DAL_EXPORT descriptor_base : public base {
...
     using method_t = Method;
 
+    explicit descriptor(std::int64_t class_count, std::int64_t neighbor_count) {
+        set_class_count(class_count);
+        set_neighbor_count(neighbor_count);
+    }
+
     auto& set_class_count(std::int64_t value) {
-        set_class_count_impl(value);
+        descriptor_base<Task>::set_class_count_impl(value);
```
Comment (mchernov-intel) [inline cpp/oneapi/dal/algo/knn/common.hpp:80]: same

File: cpp/oneapi/dal/algo/knn/common.hpp [hunk 10]
```diff
@@ -48,36 +59,44 @@ class ONEAPI_DAL_EXPORT descriptor_base : public base {
...
         return *this;
     }
 
     auto& set_neighbor_count(std::int64_t value) {
-        set_neighbor_count_impl(value);
+        descriptor_base<Task>::set_neighbor_count_impl(value);
```
Comment (mchernov-intel) [inline cpp/oneapi/dal/algo/knn/common.hpp:85]: same

**Review by KulikovNikita — APPROVED** [review]

**Review by KulikovNikita — APPROVED** [review]

**Review by rlnx — APPROVED** [review]

**Pull request merged**